### PR TITLE
[JENKINS-69817] Fix addition of SubTaskTimeInQueueActions

### PR DESCRIPTION
**Pull request by Dohbedoh**
[JENKINS-69817](https://issues.jenkins.io/browse/JENKINS-69817)

* Supersedes https://github.com/jenkinsci/metrics-plugin/pull/190 with tests. cc @sundergopalsingh
* Fixes a regression caused by https://github.com/jenkinsci/metrics-plugin/pull/127 using the same logic that was there, that is find the main owner using a `while` loop and compare it to the current completed task: 

```
            Queue.Task owner = task.getOwnerTask();
            while (owner != owner.getOwnerTask()) {
                owner = owner.getOwnerTask();
            }
            if (owner != task) {
```

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by Dohbedoh** [comment]
ping @jenkinsci/metrics-plugin-developers

**Review by raul-arabaolaza — APPROVED** [review]

**Comment by julieheard** [comment]
@stephenc @alecharp
Would it be possible to get this merged?

File: pom.xml [hunk 1]
```diff
@@ -62,7 +62,7 @@
   </scm>
 
   <properties>
-    <jenkins.version>2.346.1</jenkins.version>
+    <jenkins.version>2.346.3</jenkins.version>
```
Comment (alecharp) [inline pom.xml:65]: Any reason for this specific upgrade?

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java [hunk 2]
```diff
@@ -136,6 +143,34 @@ public void given__a_job__when__built__then__events_and_metrics_include_build()
...
+    @Test
+    @Issue("JENKINS-69817")
+    public void given__a_job_with_subtasks__when__built__then__build_include_subtasks_metrics() throws Exception {
+        MyListener listener = ExtensionList.lookup(QueueItemMetricsListener.class).get(MyListener.class);
+        assertThat(listener, notNullValue());
+        
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "testSubTasks");
+        p.setDefinition(new CpsFlowDefinition("node { sleep 1 }" +
+            "\nnode { sleep 2 }", true));
+//        j.jenkins.setQuietPeriod(0);
```
Comment (alecharp) [inline src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:155]:
```suggestion
```

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java [hunk 3]
```diff
@@ -136,6 +143,34 @@ public void given__a_job__when__built__then__events_and_metrics_include_build()
...
+        List<TimeInQueueAction> timeInQueueActions = workflowRun.getActions(TimeInQueueAction.class);
+        assertThat(timeInQueueActions, notNullValue());
+        assertThat(timeInQueueActions, hasSize(1));
+        
+        List<SubTaskTimeInQueueAction> subTaskActions = workflowRun.getActions(SubTaskTimeInQueueAction.class);
+        assertThat(subTaskActions, notNullValue());
+        assertThat(subTaskActions, hasSize(2));
+        assertThat(subTaskActions.get(0).getExecutingDurationMillis()/1000, equalTo(1L));
+        assertThat(subTaskActions.get(1).getExecutingDurationMillis()/1000, equalTo(2L));
+
```
Comment (alecharp) [inline src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:171]:
```suggestion
```

**Review by alecharp — COMMENTED** [review]
I'm not seeing any error, but wonder why the minor update on core?

File: pom.xml (hunk 1 above)
Comment (jglick) [reply pom.xml:65]: https://www.jenkins.io/doc/developer/plugin-development/choosing-jenkins-baseline/ but does not belong in this PR

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java [hunk 4]
```diff
@@ -136,6 +143,34 @@ public void given__a_job__when__built__then__events_and_metrics_include_build()
         assertThat(events.get(2).getExecutorCount().orElse(null), is(1));
     }
 
+    @Test
+    @Issue("JENKINS-69817")
+    public void given__a_job_with_subtasks__when__built__then__build_include_subtasks_metrics() throws Exception {
+        MyListener listener = ExtensionList.lookup(QueueItemMetricsListener.class).get(MyListener.class);
+        assertThat(listener, notNullValue());
```
Comment (jglick) [inline src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:150]:
```suggestion
        MyListener listener = ExtensionList.lookupSingleton(MyListener.class);
```

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java [hunk 5]
```diff
@@ -136,6 +143,34 @@ public void given__a_job__when__built__then__events_and_metrics_include_build()
...
+        
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "testSubTasks");
```
Comment (jglick) [inline src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:152]:
```suggestion
        WorkflowJob p = j.createProject(WorkflowJob.class);
```

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java [hunk 6]
```diff
@@ -136,6 +143,34 @@ public void given__a_job__when__built__then__events_and_metrics_include_build()
...
+        p.setDefinition(new CpsFlowDefinition("node { sleep 1 }" +
+            "\nnode { sleep 2 }", true));
+//        j.jenkins.setQuietPeriod(0);
+        WorkflowRun workflowRun = Optional.ofNullable(p.scheduleBuild2(0))
+            .orElseThrow(AssertionFailedError::new)
+            .waitForStart();
+        j.waitForCompletion(workflowRun);
+        j.waitUntilNoActivity();
```
Comment (jglick) [inline src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:160]:
```suggestion
        WorkflowRun workflowRun = j.buildAndAssertAccess(p);
```

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java [hunk 7]
```diff
@@ -136,6 +143,34 @@ public void given__a_job__when__built__then__events_and_metrics_include_build()
...
+
+        List<TimeInQueueAction> timeInQueueActions = workflowRun.getActions(TimeInQueueAction.class);
+        assertThat(timeInQueueActions, notNullValue());
+        assertThat(timeInQueueActions, hasSize(1));
+        
+        List<SubTaskTimeInQueueAction> subTaskActions = workflowRun.getActions(SubTaskTimeInQueueAction.class);
+        assertThat(subTaskActions, notNullValue());
```
Comment (jglick) [inline src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:167]: Is this what fails with the `src/main/` change?

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java [hunk 8]
```diff
@@ -136,6 +143,34 @@ public void given__a_job__when__built__then__events_and_metrics_include_build()
...
+        assertThat(subTaskActions, hasSize(2));
+        assertThat(subTaskActions.get(0).getExecutingDurationMillis()/1000, equalTo(1L));
+        assertThat(subTaskActions.get(1).getExecutingDurationMillis()/1000, equalTo(2L));
```
Comment (jglick) [inline src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:170]: Smells flaky. https://hamcrest.org/JavaHamcrest/javadoc/2.2/org/hamcrest/number/IsCloseTo.html ?

**Review by jglick — APPROVED** [review]
Thanks!

**Comment by raul-arabaolaza** [comment]
@Dohbedoh I have fixed the merge conflicts, lets see if this passes the build, also if you want/can take into account the non blocking feedback

File: pom.xml (hunk 1 above)
Comment (alecharp) [reply pom.xml:65]: sure, about choosing, but for this PR I don't know what the plugin would benefit from the `2.346.3` that the `2.346.1` is not providing..

Comment (Dohbedoh) [reply pom.xml:65]: @alecharp  required for the `workflow-*` test dependencies.

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java (hunk 8 above)
Comment (Dohbedoh) [reply src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:170]: Yeah, was wondering about this. I ran the test multiple times and was never failing. But I guess using `isCloseTo` might be better.

File: src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java (hunk 7 above)
Comment (Dohbedoh) [reply src/test/java/jenkins/metrics/impl/JenkinsMetricProviderImplTest.java:157]: The current problem is that `SubTaskTimeInQueueAction` sare never added to the builds. The main test that validate the fix is:

```
assertThat(subTaskActions, hasSize(2));
```

**Review by raul-arabaolaza — APPROVED** [review]
Feedback has been addressed and the pr was previously approved before, so merging

**Pull request merged**